main/wifi.c

```diff
@@ -77,14 +78,60 @@ static esp_err_t event_handler(void *ctx, system_event_t *event)
     return ESP_OK;
 }
 
-int wifi_connect(const char *ssid, const char *password)
+eap_method_t wifi_eap_atomethod(const char *method)
 {
+    if (method == NULL)
+        return EAP_NONE;
+
+    struct {
+        const char *name;
+        int method;
+    } *p, methods[] = {
+        { "TLS", EAP_TLS },
+        { "PEAP", EAP_PEAP },
+        { "TTLS", EAP_TTLS },
+        { NULL, EAP_NONE }
+    };
+
+    for (p = methods; p->name; p++)
+    {
+        if (!strcmp(p->name, method))
+            break;
+    }
+
+    return p->method;
+}
+
+int wifi_connect(const char *ssid, const char *password,
+    eap_method_t eap_method, const char *eap_identity,
+    const char *eap_username, const char *eap_password,
+    const char *ca_cert, const char *client_cert, const char *client_key)
+{
+    esp_wpa2_config_t config = WPA2_CONFIG_INIT_DEFAULT();
     wifi_config_t wifi_config = {};
     strncpy((char *)wifi_config.sta.ssid, ssid, 32);
-    strncpy((char *)wifi_config.sta.password, password, 64);
+    if (password)
+        strncpy((char *)wifi_config.sta.password, password, 64);
 
     ESP_ERROR_CHECK(esp_wifi_set_mode(WIFI_MODE_STA));
     ESP_ERROR_CHECK(esp_wifi_set_config(ESP_IF_WIFI_STA, &wifi_config));
+    if (eap_method) {
+        ESP_ERROR_CHECK(esp_wifi_sta_wpa2_ent_set_ca_cert((uint8_t *)ca_cert,
+            ca_cert ? strlen(ca_cert) : 0));
+        ESP_ERROR_CHECK(esp_wifi_sta_wpa2_ent_set_cert_key((uint8_t *)client_cert,
```

Comment:
Why not check for `client_cert && client_key` before calling `esp_wifi_sta_wpa2_ent_set_cert_key()`?
I think it will be a bit more readable. Same for line 119 above.

Reply:
I will check for `client_cert` before calling `esp_wifi_sta_wpa2_ent_set_cert_key()`. However, `client_key` is optional as it is used when the key from the client certificate is encrypted.

Should I check for `client_key` and write two different calls to `esp_wifi_sta_wpa2_ent_set_cert_key()` considering if `client_key` is present or not?

Another option is to drop the feature entirely since it is possible to put the decrypted key in the certificate directly.

Reply:
> `client_key` is optional as it is used when the key from the client certificate is encrypted.

Are you sure about that? This should be the private key corresponding with the public key set in the certificate, not a passphrase used to protect the private key. I don't think this is optional at all.

Reply:
> Are you sure about that? This should be the private key corresponding with the public key set in the certificate, not a passphrase used to protect the private key.

You are indeed right as the `private_key_password` is another field: https://github.com/espressif/esp-idf/blob/7dd49231903643d641ed4ccaff01422a9db51fb2/components/esp32/include/esp_wpa2.h#L167-L172.

> I don't think this is optional at all.

I would expect it to be optional if the private key is included in the client certificate. E.g., [wpa2_client.pem](https://github.com/espressif/esp-idf/blob/release/v3.2/examples/wifi/wpa2_enterprise/main/wpa2_client.pem) from the example contains the private key (which is encrypted—denoted with `BEGIN ENCRYPTED PRIVATE KEY` or `Proc-Type: 4,ENCRYPTED` metadata). Whenever the ESP32 extract the private key from the client certificate will remain a mystery until tested or asking Espressif since the underlying implementation is a binary blob (same is true for [SSL](https://github.com/espressif/esp-idf/blob/7dd49231903643d641ed4ccaff01422a9db51fb2/components/tcp_transport/transport_ssl.c#L172-L188)).

Reply:
I asked the question on [ESP32 Forum](https://www.esp32.com/viewtopic.php?f=13&t=12233) since that it might be useful to other people and testing it is cumbersome (especially for EAP).

Reply:
Thanks for checking that out. In their example, they still pass a key file for the connection so it does seem odd.
If it's OK by you, I would rather wait for an answer on the forum before merging.